src/serial.cpp

```diff
@@ -0,0 +1,332 @@
+#include "serial.h"
+#include "status_codes.h"
+
+#include <cstring>
+#include <dirent.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/select.h>
+#include <termios.h>
+#include <unistd.h>
+#include <utility>
+
+// -----------------------------------------------------------------------------
+// Global callback function pointers (default nullptr)
+// -----------------------------------------------------------------------------
+void (*errorCallback)(int) = nullptr;
+void (*readCallback)(int) = nullptr;
+void (*writeCallback)(int) = nullptr;
+
+// -----------------------------------------------------------------------------
+// Internal helpers & types
+// -----------------------------------------------------------------------------
+namespace {
+
+struct SerialPortHandle {
+  int fd;
+  termios original; // keep original settings so we can restore on close
+};
+
+// Map integer baudrate to POSIX speed_t. Only common rates are supported.
+static auto to_speed_t(int baud) -> speed_t {
+  switch (baud) {
+  case 0:
+    return B0;
+  case 50:
+    return B50;
+  case 75:
+    return B75;
+  case 110:
+    return B110;
+  case 134:
+    return B134;
+  case 150:
+    return B150;
+  case 200:
+    return B200;
+  case 300:
+    return B300;
+  case 600:
+    return B600;
+  case 1200:
+    return B1200;
+  case 1800:
+    return B1800;
+  case 2400:
+    return B2400;
+  case 4800:
+    return B4800;
+  case 9600:
+    return B9600;
+  case 19200:
+    return B19200;
+  case 38400:
+    return B38400;
+  case 57600:
+    return B57600;
+  case 115200:
+    return B115200;
+  case 230400:
+    return B230400;
+#ifdef B460800
+  case 460800:
+    return B460800;
+#endif
+#ifdef B921600
+  case 921600:
+    return B921600;
+#endif
+  default:
+    return B9600; // fallback
+  }
+}
+
+inline void invokeError(int code) {
+  if (errorCallback) {
+    errorCallback(code);
+  }
+}
+
+} // namespace
+
+// -----------------------------------------------------------------------------
+// Public API implementation
+// -----------------------------------------------------------------------------
+
+intptr_t serialOpen(void *port, int baudrate, int dataBits, int parity,
+                    int stopBits) {
```

Comment:
This parameter wrapping is ugly as hell. Please change the clang-tidy config so that if there are to many parameters they overflow like this:

```suggestion
intptr_t serialOpen(
  void *port,
  int baudrate,
  int dataBits,
  int parity,
  int stopBits
) {
```